src/main/java/com/app/onlinebookstore/service/impl/UserServiceImpl.java

```diff
@@ -0,0 +1,52 @@
+package com.app.onlinebookstore.service.impl;
+
+import com.app.onlinebookstore.dto.UserRegistrationRequestDto;
+import com.app.onlinebookstore.dto.UserRegistrationResponseDto;
+import com.app.onlinebookstore.exaption.RegistrationException;
+import com.app.onlinebookstore.mapper.UserMapper;
+import com.app.onlinebookstore.model.Role;
+import com.app.onlinebookstore.model.RoleName;
+import com.app.onlinebookstore.model.User;
+import com.app.onlinebookstore.repository.RoleRepository;
+import com.app.onlinebookstore.repository.UserRepository;
+import com.app.onlinebookstore.service.UserService;
+import java.util.HashSet;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final RoleRepository roleRepository;
+    private final PasswordEncoder passwordEncoder;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserRegistrationResponseDto register(
+            UserRegistrationRequestDto request)
+            throws RegistrationException {
+        if (userRepository.findByEmail(
+                request.getEmail()).isPresent()) {
+            throw new RegistrationException("Unable to complete registration.");
+        }
+        User user = new User();

```

Comment:
Optional: implement toModel method in mapper, for more readable code (all stuff such as password encode do in in service)